src/shared/callbacks.rs

```diff
@@ -0,0 +1,147 @@
+use js_sys::{Function, Promise};
+use wasm_bindgen::prelude::*;
+use wasm_bindgen_futures::{spawn_local, JsFuture};
+
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+
+#[derive(Default)]
+pub struct ResolvableState {
+    pub waker: Option<Waker>,
+    pub result: Vec<String>,
+}
+
+#[derive(Clone)]
+pub struct Resolvable {
+    pub resolved: Arc<Mutex<bool>>,
+    pub state: Arc<Mutex<ResolvableState>>,
+}
+
+#[allow(clippy::mutex_atomic)]
+impl Default for Resolvable {
+    fn default() -> Self {
+        Resolvable {
+            resolved: Arc::new(Mutex::new(false)),
+            state: Arc::new(Mutex::new(ResolvableState::default())),
+        }
+    }
+}
+
+impl Resolvable {
+    pub fn resolve(&mut self, val: Vec<String>) {
+        if let Ok(mut resolved) = self.resolved.lock() {
+            *resolved = true;
+        }
+
+        if let Ok(mut state) = self.state.lock() {
+            if let Some(waker) = state.waker.clone() {
+                waker.wake();
+            }
+
+            state.result = val;
+        }
+    }
+
+    pub fn result(&mut self) -> Vec<String> {
+        if let Ok(state) = self.state.lock() {
+            return state.result.clone();
+        }
+
+        vec![]
+    }
+
+    pub fn is_resolved(&self) -> bool {
+        if let Ok(resolved) = self.resolved.lock() {
+            return *resolved;
+        }
+
+        false
+    }
+}
+
+impl Future for Resolvable {
+    type Output = ();
+
+    fn poll(
+        self: Pin<&mut Self>,
+        ctx: &mut Context,
+    ) -> Poll<Self::Output> {
+        if self.is_resolved() {
+            Poll::Ready(())
+        } else {
+            let waker = ctx.waker().clone();
+            if let Ok(mut state) = self.state.lock() {
+                state.waker = Some(waker);
+            }
+            Poll::Pending
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Callback {
+    f: Function,
+}
+
+impl Callback {
+    pub fn new(f: Function) -> Self {
+        Callback { f }
+    }
+
+    pub fn call0(&self) -> Result<Promise, String> {
+        match self.f.call0(&JsValue::NULL) {
+            Ok(result) => Ok(Promise::from(result)),
+            Err(_e) => Err("Callback failed".to_string()),
+        }
+    }
+}
+
+unsafe impl<'a> Send for Callback {}
```

Comment:
can you help me understand what the unsafe usage is for?

Reply:
since it is async, data types must be able to cross threads safely but Function from js_sys (which is a property of Callback) is not thread safe for most use cases. However in our case it is fine and this just says that I explicitly say Callback is safe to cross boundaries.

Reply:
Dumb question, but is there any way to make it safe, or is it just unavoidable because of the way that it's boundary crossing between js?

Reply:
JS makes this unsafe, Callback contains Function which contains a u8 pointer (which is the problem) and Function is part of the js-sys crate